### PR TITLE
Local page saving

### PR DESCRIPTION
**Pull request by thehabbos007**
## 🌟 What is the purpose of this PR?

Introducing the Graph Layer left pages in a broken state. This PR (1) reintroduces "local" pages that are loaded and saved without collab, and (2) moves some of the shared collab functionality to the front-end, changing how the frontend interacts with the document. Resultantly, text entities are no longer separate from the block's "data entity".

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1202953520344821/1203132467619201/f) _(internal)_

## 🔍 What does this change?

- Disables collab and its position reporters
- Changes the entity store to work with the new type system
- Removes legacy link usage on the frontend
- Fixes page saving actions on the backend to work with the frontend request format
- Re-add a save ProseMirror plugin to save pages (with debouncing)
- Changes typings in the prosemirror manager/entity store
- Moves most page-related operations new GQL resolvers/types

## ⚠️ Known issues

- There are still some issues on the page that may appear as the entity store may not be fully compatible with the new setup
- Blocks are still breaking as the API we prove to them do not really work with the backend
- Variants may be broken under some circumstances because of saving entities without the new graph layer's property scheme (also variant properties are stored in-line with text data now)
- No realtime updates on pages requiring manual reloading to stay up to date.
- The operation ordering of updating the entity store may still be somewhat broken
- Untested features of the frontend which are untouched (but passes type-checks)
- Selection removal is somewhat buggy
- Saving isn't instantaneous, and leaving the page preemptively may lead to lost progress
- Typing a little too fast may also lead to some lost progress

## 🐾 Next steps

- Collab v2 using Yjs and the new Graph Layer natively

## 🛡 What tests cover this?

- None :(

## ❓ How to test this?

Check out the branch, start external services and `yarn dev`. Create any page and start typing/editing/adding blocks

## 📹 Demo

(network error after reload is due to stopping an active request)
https://user-images.githubusercontent.com/6988668/195583514-301970d6-3fd2-4394-89b2-15d462207a83.mp4

### CODE REVIEW

**Comment by nonparibus** [comment]
Even if there are errors/details to improve on in this we shouldn't focus too much on them rn. Priority is to merge in, and this will change anyway with new collab. A few of the methods might survive, but saving itself will be handled by the new collab server.

File: packages/hash/shared/src/entityStorePlugin.ts [hunk 1]
```diff
@@ -64,6 +59,10 @@ export type EntityStorePluginAction = { received?: boolean } & (
       type: "updateEntityProperties";
       payload: {
         draftId: string;
+        metadata?: {
+          componentId: string;
+          dataEntity?: { [key: string]: unknown };
+        };
         properties: { [key: string]: unknown };
         merge: boolean;
       };
```
Comment (CiaranMn) [inline packages/hash/shared/src/entityStorePlugin.ts:68]: Assuming it should only ever be one or the other.
```suggestion
      payload: {
        draftId: string;
        merge: boolean;
      } & (
        | {
            metadata?: {
              componentId: string;
              dataEntity?: { [key: string]: unknown };
            };
          }
        | { properties: { [key: string]: unknown } }
      );
    }
```

 I've commented elsewhere that we could make it clearer via naming where we're introducing special provision for entities referenced by blocks, which might result in something like:
```suggestion
      payload: {
        draftId: string;
        merge: boolean;
      } & (
        | {
            blockEntityMetadata?: {
              componentId: string;
              blockChildEntity?: { [key: string]: unknown };
            };
          }
        | { properties: { [key: string]: unknown } }
      );
    }
```

File: packages/hash/shared/src/queries/page.queries.ts [hunk 2]
```diff
@@ -1,5 +1,40 @@
 import { gql } from "@apollo/client";
 
+const persistedBlockFieldsFragment = gql`
+  fragment PersistedBlockFields on PersistedBlock {
+    __typename
+    entityId
+    entityVersion
+    accountId
+    entityTypeId
+    componentId
```
Comment (CiaranMn) [inline packages/hash/shared/src/queries/page.queries.ts:10]: I think `componentId` should be in `properties` – a block is an entity like any other, which in its pure BP representation should look like
```
{
  entityId
  entityTypeId
  properties {
    componentId
  }
}
```
with a separate link to its child entity

**Review by CiaranMn — COMMENTED** [review]
Thanks @thehabbos007 – exciting to see everything coming back together again

I have left a couple of comments which are around two themes:
1. I want to make sure we call out where we are maintaining special abstractions / fakery for block entities, which could be renaming variables in this PR. And make sure we look for opportunities remove / minimise them in future PRs
2. I'm not sure about the shape of `PersistedBlockFields` - I would've thought `componentId` should be in its `properties`, since it's one of its actual owned properties stored on it in the db (unlike the child entity, which is stored as a link, and has a special convenience field here – `dataEntity` – which we've discussed removing in favour of the FE reconstructing trees from linked entities as it needs. I appreciate this cleanup is out of the scope of this PR.)

**Comment by nathggns** [comment]
Reviewing this now

File: packages/hash/frontend/src/blocks/page/ComponentView.tsx [hunk 3]
```diff
@@ -42,13 +42,13 @@ const BLANK_PROPERTIES = {};
 
 const getChildEntity = (
   entity: DraftEntity | null | undefined,
-): DraftEntity<BlockEntity["properties"]["entity"]> | null => {
-  if (entity) {
+): DraftEntity<BlockEntity["dataEntity"]> | null => {
```
[the remark this answers is not in the document]
Comment (nathggns) [reply packages/hash/frontend/src/blocks/page/ComponentView.tsx:45]: I've asked about this before, but it seems like the backend team are happy with these kind of "named" shortcut resolvers for links. It would be good to agree a collective approach. https://github.com/hashintel/hash/pull/1060#discussion_r971044213

[the remark this answers is not in the document]
Comment (CiaranMn) [reply packages/hash/frontend/src/blocks/page/ComponentView.tsx:45]: We had previously agreed with this sentiment:
> I think we're probably better off just using links and having some functions clientside that construct whatever trees we want, to make the resolution of data less mysterious

([internal Slack link](https://hashintel.slack.com/archives/C03F7V6DU9M/p1661177030351259?thread_ts=1661176442.884289&cid=C03F7V6DU9M), no further context except agreement)

I am not firmly opposed to these helpers, but I think their existence encourages the abstraction to bleed further across the stack than it needs to. I _am_ firmly opposed (absent any good argument, which might come up) for it being something that's handled in the save process, because I think it's introducing special casing that is not required and makes the whole system harder to understand and maintain.

File: packages/hash/frontend/src/blocks/page/Comments/CommentTextField.tsx [hunk 4]
```diff
@@ -75,11 +77,10 @@ export const CommentTextField: FunctionComponent<CommentTextFieldProps> = ({
           keymap<Schema>({
             Enter(_, __, view) {
               if (!loadingRef.current && view?.state.doc.content) {
-                const { tokens } = textBlockNodeToEntityProperties(
-                  view.state.doc,
-                );
+                const { [TEXT_TOKEN_PROPERTY_TYPE_BASE_URI]: tokens } =
+                  textBlockNodeToEntityProperties(view.state.doc);
```
Comment (nathggns) [inline packages/hash/frontend/src/blocks/page/Comments/CommentTextField.tsx:81]:
```suggestion
                const tokens =
                  textBlockNodeToEntityProperties(view.state.doc)[TEXT_TOKEN_PROPERTY_TYPE_BASE_URI];
```

File: packages/hash/frontend/src/blocks/page/ComponentView.tsx [hunk 5]
```diff
@@ -42,13 +42,13 @@ const BLANK_PROPERTIES = {};
...
+  if (entity && entity.dataEntity) {
     if (!isDraftBlockEntity(entity)) {
       throw new Error("Cannot prepare non-block entity for prosemirrior");
     }
 
-    return entity.properties.entity;
+    return entity.dataEntity as DraftEntity;
```
Comment (nathggns) [inline packages/hash/frontend/src/blocks/page/ComponentView.tsx:51]: What is this casting needed for?

File: packages/hash/frontend/src/blocks/page/createEditorView.ts [hunk 6]
```diff
@@ -21,6 +31,85 @@ import { createTextEditorView } from "./createTextEditorView";
...
+  );
+
+  return new Plugin<unknown, Schema>({
+    view: (_viewOnCreation: EditorView<Schema>) => {
+      return {
+        update: (view, prevState) => {
+          if (view.state.doc !== prevState.doc) {
+            // If the document changes between updates, we issue a debounced update.
+            writeDebounce(view);
+          }
```
Comment (nathggns) [inline packages/hash/frontend/src/blocks/page/createEditorView.ts:83]: What about if the entity store changes?

File: packages/hash/frontend/src/blocks/page/createEditorView.ts [hunk 7]
```diff
@@ -105,6 +196,14 @@ export const createEditorView = (
     },
   );
 
+  void manager.loadPage(state, getLastSavedValue()).then((tr) => {
+    view.updateState(state.apply(tr));
```
Comment (nathggns) [inline packages/hash/frontend/src/blocks/page/createEditorView.ts:200]: Is there a reason we're doing this over using `view.dispatch`?

File: packages/hash/shared/src/queries/page.queries.ts (hunk 2 above)
Comment (benwerner01) [reply packages/hash/shared/src/queries/page.queries.ts:10]: It is available in the properties, but the key would be its base URI so that query would not work. The top-level `componentId` field resolver acts like an alias for that property.

File: packages/hash/shared/src/save.ts [hunk 8]
```diff
@@ -483,13 +349,19 @@ const calculateSaveActions = async (
       beforeBlockDraftIds.splice(position, 0, afterDraftId);
     }
   }
+
+  actions.push(
+    ...moveActions.filter((action) => action.removeBlock),
+    ...moveActions.filter((action) => action.insertBlock),
+  );
```
Comment (nathggns) [inline packages/hash/shared/src/save.ts:356]: Why is this change necessary? Is this fixing a previously unknown bug?

File: packages/hash/shared/src/queries/page.queries.ts (hunk 2 above)
Comment (CiaranMn) [reply packages/hash/shared/src/queries/page.queries.ts:10]: Ah right, I understand the motivation now, I wasn't thinking about the fact that this was now a URI.

I don't think adding a field resolver when we want to access properties on a type defined in the workspace API is a good pattern to establish. It adds unnecessary overhead, and makes the presence of the `properties` blob pointless since we're never actually using it. It also means that the FE needs to know to translate `componentId` back into its URI when sending a properties _update_, or (even worse in my view) we need to then add special handlers for every property when they are updated to allow the FE to use this shorter form.

Comment (CiaranMn) [reply packages/hash/shared/src/queries/page.queries.ts:10]: It also makes it unclear which of these things are truly top-level fields on an entity, which are special handlers for other entities (`dataEntity`), and which are shortcuts to things in the `properties` blob.

The amount of special casing and 'this thing isn't really there but we pretend it is' makes the current system very confusing, and I'd rather we kept it to a minimum in the new one.

File: packages/hash/shared/src/ProsemirrorManager.ts [hunk 9]
```diff
@@ -323,49 +315,20 @@ export class ProsemirrorManager {
...
       blockEntity &&
-      areComponentsCompatible(
-        /** @todo this any type coercion is incorrect, we need to adjust typings https://app.asana.com/0/0/1203099452204542/f */
-        (blockEntity as any).componentId,
-        targetComponentId,
-      )
+      areComponentsCompatible(blockEntity.componentId, targetComponentId)
     ) {
-      if (targetComponentId === blockEntity?.properties.componentId) {
+      if (targetComponentId === blockEntity.componentId) {
```
[the remark this answers is not in the document]
Comment (CiaranMn) [reply packages/hash/shared/src/ProsemirrorManager.ts:320]: I suggest we keep further discussion on this point to this similar comment, so I will resolve this one: https://github.com/hashintel/hash/pull/1196/files/4db20e72149725297f5560645c1907d4ca862769#r997268751

File: packages/hash/shared/src/queries/page.queries.ts (hunk 2 above)
Comment (CiaranMn) [reply packages/hash/shared/src/queries/page.queries.ts:10]: ([reference to other comment](https://github.com/hashintel/hash/pull/1196/files/4db20e72149725297f5560645c1907d4ca862769#r997153374) this point applies to – I've resolved it to keep discussion in one place)

File: packages/hash/frontend/src/components/BlockLoader/BlockLoader.tsx [hunk 10]
```diff
@@ -72,11 +68,8 @@ type BlockLoaderProps = {
 
 const removeTextEntities = (properties: {}) => {
   // @see https://app.asana.com/0/1201095311341924/1202694273052398/f
+  // NOTE, THIS HAS BEEN CHANGED SINCE to accompany the new Graph Layer
```
[the remark this answers is not in the document]
Comment (thehabbos007) [reply packages/hash/frontend/src/components/BlockLoader/BlockLoader.tsx]: I've removed `removeTextEntities` completely, as I think it was causing some rendering issues as we are linking directly to text entities now
fe49107ff0fd2893676d5efbe5db2d54543980cf

File: packages/hash/shared/src/ProsemirrorManager.ts [hunk 11]
```diff
@@ -602,11 +528,12 @@ export class ProsemirrorManager {
       payload: {
         draftId: newBlockId,
         merge: false,
-        properties: {
+        metadata: {
```
[the remark this answers is not in the document]
Comment (thehabbos007) [reply packages/hash/shared/src/ProsemirrorManager.ts]: Yes, exactly. 
I've renamed the field in ed28f4823179c6d1a42c59b433c1701e8a565688

File: packages/hash/shared/src/entityStorePlugin.ts (hunk 1 above)
Comment (thehabbos007) [reply packages/hash/shared/src/entityStorePlugin.ts]: It shouldn't always be one or the other, the metadata can be set at the same time as the properties

File: packages/hash/shared/src/queries/page.queries.ts (hunk 2 above)
Comment (thehabbos007) [reply packages/hash/shared/src/queries/page.queries.ts:10]: This is currently done for workspace types as we have full control, but I see how it could be confusing/misleading when dealing with `block`s as the API is somewhat exposed to the underlying "components". We could move this back out to properties and make use of the ontology URIs from the shared project to access the property field, but I'm not sure we would gain much if this code is going to be thrown away relatively soon.

I think it's worth it to make the change for new collab, though, as we do have the opportunity to make things more consistent.

File: packages/hash/shared/src/entityStorePlugin.ts (hunk 1 above)
Comment (CiaranMn) [reply packages/hash/shared/src/entityStorePlugin.ts]: A block entity only has `componentId` as a property, though (and a link to another entity). And no other entity can use `blockEntityMetadata`. What is the use case for setting both at once?

File: packages/hash/shared/src/entity.ts [hunk 12]
```diff
@@ -5,110 +5,51 @@ import {
...
+};
+
 // @todo make this more robust
 export const isTextProperties =
-  (properties: {}): properties is TextProperties => "tokens" in properties;
+  (properties: {}): properties is TextEntityType["properties"] =>
+    TEXT_TOKEN_PROPERTY_TYPE_BASE_URI in properties;
 
 // @todo make this more robust, checking system type name of entity type
 export const isTextEntity = (
```
[the remark this answers is not in the document]
Comment (thehabbos007) [reply packages/hash/shared/src/entity.ts]: Turns out we can't do this as is because draft entities use this check. Had to remove again in 
b6eec1eabd5440c1aaccf4f979b7e1c98daf2b87

[the remark this answers is not in the document]
Comment (CiaranMn) [reply packages/hash/shared/src/entity.ts]: Fair enough – when we can introduce it, won't it need to account for versioning? Unless the version of the system type never changes. Not important for this PR if we aren't introducing it

File: packages/hash/shared/src/queries/page.queries.ts (hunk 2 above)
Comment (CiaranMn) [reply packages/hash/shared/src/queries/page.queries.ts:10]: Sounds good – presumably the FE will be going through collab anyway, and collab will be the consumer of these types.

File: packages/hash/shared/src/entityStorePlugin.ts (hunk 1 above)
Comment (thehabbos007) [reply packages/hash/shared/src/entityStorePlugin.ts]: My bad, i was looking at an instance where I had forgoten to remove the properties where I was using `blockEntityMetadata`. I'll update this soon!

Comment (thehabbos007) [reply packages/hash/shared/src/entityStorePlugin.ts]: Addressed in 0087f8337a0b6cc0badb201a7b3e6c4a45710447
will rename field next

File: packages/hash/shared/src/entity.ts [hunk 13]
```diff
@@ -5,110 +5,51 @@ import {
...
-  return (
-    isTextContainingEntityProperties(entityProperties) &&
-    isDraftEntity(entityProperties.text.data)
-  );
-};
-
-/**
- * @todo this will need to change when we remove legacy links
- */
 export const getEntityChildEntity = (
```
[the remark this answers is not in the document]
Comment (thehabbos007) [reply packages/hash/shared/src/entity.ts:56]: Yes, fair point. I'm not sure how easy this is to refactor right now, as the names overlap a little. This `getEntityChildEntity` is the `getBlockChildEntity` implementation, but now that we're linking directly to the data, it doesn't make much sense anymore.

File: packages/hash/shared/src/queries/page.queries.ts (hunk 2 above)
Comment (thehabbos007) [reply packages/hash/shared/src/queries/page.queries.ts:10]: Just for clarification, do you want me to move this out to `properties` again, or should we skip this?

File: packages/hash/frontend/src/blocks/page/createEditorView.ts (hunk 6 above)
Comment (thehabbos007) [reply packages/hash/frontend/src/blocks/page/createEditorView.ts]: Debounce + interval saving in b6eec1eabd5440c1aaccf4f979b7e1c98daf2b87 (seems to be a little more stable with this approach)

File: packages/hash/frontend/src/blocks/page/createEditorView.ts (hunk 7 above)
Comment (thehabbos007) [reply packages/hash/frontend/src/blocks/page/createEditorView.ts:218]: No reason. Any advantages to `dispatch`? I would be interesting to know if I should be doing something else in 
https://github.com/hashintel/hash/blob/4db20e72149725297f5560645c1907d4ca862769/packages/hash/frontend/src/blocks/page/createEditorView.ts#L60
as well.

From what I understand, we wouldn't be touching anything other than the `state` prop here, so I chose to not dispatch

File: packages/hash/shared/src/save.ts (hunk 8 above)
Comment (thehabbos007) [reply packages/hash/shared/src/save.ts:356]: Yes, it's about dealing with the lack of atomicity. The backend does not play well with inserting a block at a position before removal

File: packages/hash/api/codegen.yml [hunk 14]
```diff
@@ -11,7 +11,7 @@ generates:
   ./src/graphql/apiTypes.gen.ts:
     documents:
       - ../shared/src/queries/**/*.ts
-      - ./src/collab/graphql/queries/**/*.ts
+      # - ./src/collab/graphql/queries/**/*.ts
```
[the remark this answers is not in the document]
Comment (thehabbos007) [reply packages/hash/api/codegen.yml]: No plan on adding it back right now. 
Removed in c28dd3bf7147b42b4bca588bd360d1523f69961a

File: packages/hash/shared/src/entityStore.ts [hunk 15]
```diff
@@ -1,31 +1,26 @@
...
-       * @deprecated
-       * @todo Don't use this, use links
-       */
-      entity: DraftEntity<Properties["entity"]>;
-    }
-  : Properties;
+export const TEXT_ENTITY_TYPE_ID = types.entityType.text.entityTypeId;
+// `extractBaseUri` does not work within this context, so this is a hacky way to get the base URI.
+export const TEXT_TOKEN_PROPERTY_TYPE_BASE_URI =
+  types.propertyType.tokens.propertyTypeId.slice(0, -3);
```
[the remark this answers is not in the document]
Comment (thehabbos007) [reply packages/hash/shared/src/entityStore.ts:13]: Unfortunately not. I've tried, but WASM is not initialized here, and I did not want to deal with injecting a dynamically imported module here

File: packages/hash/api/src/model/knowledge/page.model.ts [hunk 16]
```diff
@@ -467,17 +472,19 @@ export default class extends EntityModel {
...
       linkTypeModel: WORKSPACE_TYPES.linkType.contains,
     });
 
-    if (position < 0 || position >= contentLinks.length) {
-      throw new UserInputError(`invalid position: ${position}`);
-    }
+    // When indices are in-flux, the length is not a good enough indicator of the index.
+    // if (position < 0 || position >= contentLinks.length) {
+    //   throw new UserInputError(`invalid position: ${position}`);
+    // }
```
[the remark this answers is not in the document]
Comment (thehabbos007) [reply packages/hash/api/src/model/knowledge/page.model.ts]: Thanks, done in 77a1720b50ebe51acf8f99d6f827bac1b55f523b

File: packages/hash/api/src/graphql/resolvers/knowledge/page/update-page-contents.ts [hunk 17]
```diff
@@ -140,6 +140,7 @@ export const updatePersistedPageContents: ResolverFn<
         await pageModel.insertBlock(graphApi, {
           block: insertedBlocks[insertCount]!,
           position: action.insertBlock.position,
+          updateSiblings: false,
```
[the remark this answers is not in the document]
Comment (thehabbos007) [reply packages/hash/api/src/graphql/resolvers/knowledge/page/update-page-contents.ts:143]: Page saving produces a list of operations for all operations. If we insert item in position 2, page actions to remove all subsequent blocks are produced, followed by re-inserting them all at correct positions.

I hope that new collab will make use of server-side reordering, though.

File: packages/hash/frontend/src/blocks/page/createEditorView.ts (hunk 7 above)
Comment (thehabbos007) [reply packages/hash/frontend/src/blocks/page/createEditorView.ts]: I've updated this in 47dfa325f748dd0bde4e1541b6a19513e2de3b09

File: packages/hash/shared/src/entityStore.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (benwerner01) [reply packages/hash/shared/src/entityStore.ts:13]: What about the one that's been temporarily defined in `packages/hash/api/src/model/util.ts`?

File: packages/hash/api/src/graphql/resolvers/knowledge/page/update-page-contents.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (benwerner01) [reply packages/hash/api/src/graphql/resolvers/knowledge/page/update-page-contents.ts:143]: Thanks for explaining, hadn't appreciated that moving a block would result in all those actions being produced by collab. Agree that we need to come up with a more viable solution longterm.

Is this already captured somewhere or as part of some other task? Not sure if it's worth creating a standalone todo, just want to ensure that we don't stick to this approach for too long as it sounds very brittle.

[the remark this answers is not in the document]
Comment (thehabbos007) [reply packages/hash/api/src/graphql/resolvers/knowledge/page/update-page-contents.ts:143]: I don't think we have a place to add a task for new collab yet, but will make a note of it in the design document for new collab

**Review by CiaranMn — APPROVED** [review]

File: packages/hash/frontend/src/blocks/page/ComponentView.tsx (hunk 5 above)
Comment (nathggns) [reply packages/hash/frontend/src/blocks/page/ComponentView.tsx]: createEntityStore assumes an optional draftId because you can use it to recreate an entity store with a new set of entities but keeping the previous draft id. Don't know if this changes anything, just want to be complete.

File: packages/hash/frontend/src/blocks/page/createEditorView.ts (hunk 6 above)
Comment (nathggns) [reply packages/hash/frontend/src/blocks/page/createEditorView.ts]: You can detect entity store changes – plugin state is stored in EditorState.

File: packages/hash/frontend/src/blocks/page/createEditorView.ts (hunk 7 above)
Comment (nathggns) [reply packages/hash/frontend/src/blocks/page/createEditorView.ts]: Dispatch is standard.

File: packages/hash/api/src/graphql/resolvers/knowledge/page/update-page-contents.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (nathggns) [reply packages/hash/api/src/graphql/resolvers/knowledge/page/update-page-contents.ts:143]: This isn't the case previously, btw. Collab definitely only used to produce one action when moving a block.

File: packages/hash/frontend/src/blocks/page/createEditorView.ts (hunk 6 above)
Comment (thehabbos007) [reply packages/hash/frontend/src/blocks/page/createEditorView.ts]: I see, thank you. If this results in problems down the line, we can address it in a follow up
